### ISSUE TITLE
Zeros function not correctly determining shape.

### ISSUE THREAD

**Issue by chasep255** [tensorflow/tensorflow#2928]
I want to have dynamic batch sizes so I defined my input tensors as follows.

`input_tensor = tf.placeholder(tf.float32, (None, TIME_STEPS, 128), 'input_tensor')`

I then calculate the batch size.

`batch_size = tf.shape(input_tensor)[0]`

I can use this to determine the initial state for my RNNs as follows...

```
initial_state = state = tf.zeros((batch_size, encoder_multi_rnn.state_size), tf.float32)
...
forward_outputs[t], state = encoder_multi_rnn(getTimeStep(input_tensor, t), state)
```
This seems to work fine when I evaluate tf.shape(initial_state) in my sess.run() it prints what appears to be the correct value.  However, when do the following it prints (?, ?).  I think the second dimension should not be a ?

```
    state = tf.zeros((batch_size, encoder_multi_rnn.state_size), tf.float32)
    print(state.get_shape())
```

This is not causing an issue and my code compiles fine.  However this is causing an issue when I try to feed the output of the last time step of my sequence generating decoder back into the input of the MultiRNN.  On the first time step I feed it zeros since no output has yet to be produced.

```
    rnn_input = tf.reduce_sum(w * decoder_input, 1)
    last_out = decoder_outputs[t - 1] if t else tf.zeros((batch_size, 128))
    rnn_input = tf.concat(1, (rnn_input, last_out))
```

I get this error when concat is called.  ValueError: Linear expects shape[1] of arguments: [[None, None], [None, 1024]]

When I print last_out.get_shape() I get (?, ?) again.  In this case I think it should be (?, 128) since the last dimension is clearly defined.

**Comment by sherrym** [comment]
Hi @chasep255. If it's a runtime decision, you should do

last_out = tf.cond(t > 0, lambda: decode_output[t-1], lambda: tf.zeros((batch_size, 128))

Could you please give it a try and see if it works for you? If not, could you cut and paste the complete program? Thanks.

**Comment by chasep255** [comment]
It is not a runtime decision.  t is the time step.  Here is a more complete section.  I already fixed the issue by just defining a batch size so I don't have the original code.

Here is what is looks like now.  See how t is just the loop variable.  

```
with tf.variable_scope('decoder') as scope:
    W_attn = tf.get_variable('W_attn', (TIME_STEPS, TIME_STEPS), tf.float32, tf.truncated_normal_initializer(1.0 / TIME_STEPS, 0.01))
    W_sm = tf.get_variable('W_softmax', (decoder_gru_size, 128), tf.float32, tf.truncated_normal_initializer(0.0, 1 / np.sqrt(decoder_gru_size)))
    b_sm = tf.get_variable('b_softmax', 128, tf.float32, tf.truncated_normal_initializer(0.0, 0.01))
    decoder_outputs = [None] * TIME_STEPS
    state = tf.zeros((BATCH_SIZE, decoder_multi_rnn.state_size), tf.float32)
    for t in range(TIME_STEPS):
        w = tf.reshape(W_attn[t, :], (1, TIME_STEPS, 1))
        rnn_input = tf.reduce_sum(w * decoder_input, 1)
        last = tf.cond(is_training, lambda: getTimeStep(expected_output, t - 1), lambda: decoder_outputs[t - 1]) if t else tf.zeros((BATCH_SIZE, 128))
        rnn_input = tf.concat(1, (rnn_input, last))
        rnn_output, state = decoder_multi_rnn(rnn_input, state)
        decoder_outputs[t] = tf.nn.softmax(tf.matmul(rnn_output, W_sm) + b_sm)
        scope.reuse_variables()
    output_tensor = tf.concat(1, [tf.reshape(t, (-1, 1, 128)) for t in decoder_outputs])
```

**Comment by prb12** [comment]
Glad you managed to get it working.  I think your initial post has some confusion between the statically known shapes of ops/tensors in the graph, and the dynamically computed shapes of tensors evaluated when the graph is executed  (via session.run, feeding in placeholders) ?

e.g. The input_tensor is a placeholder with partially specified shape.  We don't know until the input is fed what the actual batch size is (and it may vary from run to run).  You then take the *dynamic* shape of the input (using a tf.shape op in the graph)  and slice off the first element.  
```
input_tensor = tf.placeholder(tf.float32, (None, TIME_STEPS, 128), 'input_tensor')
shape_of_input_tensor = tf.shape(input_tensor)
batch_size = shape_of_input_tensor[0]
```
If we examine 'shape_of_input_tensor' we see that it is a vector whose value which cannot be computed until the input is fed into the graph.
```
>>> print shape_of_input_tensor
Tensor("Shape_3:0", shape=(3,), dtype=int32)
```
Note that batch_size is actually the output of a dynamic 'Slice' operation 
```
>>> print batch_size
Tensor("Squeeze_2:0", shape=(), dtype=int32)
>>> print batch_size.op                                                                                                                                                        
name: "Squeeze"
op: "Squeeze"
input: "Slice"
attr {
  key: "T"
  value {
    type: DT_INT32
  }
}
attr {
  key: "squeeze_dims"
  value {
    list {
      i: 0
    }
  }
}
```
In contrast, 'state.get_shape()' is a static property of the graph and so cannot know the batch size until the placeholder is fed at runtime.  See here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/ops.py#L329

I couldn't reproduce further without know what 'encoder_multi_rnn.state_size' returns.  It clearly can't be a simple integer:
```
>>> STATE_SIZE=27
>>> state = tf.zeros((batch_size, STATE_SIZE), tf.float32)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 622, in zeros
    shape = ops.convert_to_tensor(shape, name="shape")
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 566, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/ops/constant_op.py", line 179, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/ops/constant_op.py", line 162, in constant
    tensor_util.make_tensor_proto(value, dtype=dtype, shape=shape))
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 332, in make_tensor_proto
    _AssertCompatible(values, dtype)
  File "/usr/local/google/home/pbar/tensorflow.virtualenv/local/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 269, in _AssertCompatible
    raise TypeError("List of Tensors when single Tensor expected")
TypeError: List of Tensors when single Tensor expected
```

Note that the args of tf.zeros are passed through [tf.convert_to_tensor](https://www.tensorflow.org/versions/r0.9/api_docs/python/framework.html#convert_to_tensor):

Without the entire original program, it's hard to follow why shape inference of last_shape (all the way through your encoder) was not working.  

Could you provide a simple repro test case?  Failing that I propose we close this issue.

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity. Please reopen when further information becomes available.
